src/main/java/org/primefaces/application/resource/csp/CspFilter.java

```diff
@@ -0,0 +1,186 @@
+ * that can then be specified via {@link Constants.ContextParams#CONTENT_SECURITY_POLICY_SUPPORTED_DIRECTIVES}.</p>
+ * <p>Per default browsers are instructed to report any CSP violations to the {@link CspReportServlet} endpoint if not
+ * provided via {@link Constants.ContextParams#CONTENT_SECURITY_POLICY_REPORT_URI}.</p>
+ * <p>To not break functionality in production you may also want to set {@link Constants.ContextParams#CONTENT_SECURITY_POLICY_REPORT_ONLY}
+ * to just report CSP violations instead of refusing them.</p>
+ * <p>If {@link Constants.ContextParams#CONTENT_SECURITY_POLICY_JAVASCRIPT_COOKIE_DEBUGGING} is enabled, it will be possible to read the CSP header
+ * in javascript with <code>document.cookie</code>. This is because the JS API unfortunately does not expose response headers. This is not recommended
+ * in production systems.</p>
+ */
+@WebFilter("/*")
+public class CspFilter implements Filter {
+
+    CspConfiguration configuration;
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        configuration = new CspConfiguration(filterConfig.getServletContext());
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        if (configuration.isEnabled()) {
+            HeaderWriterResponse headerWriterResponse = new HeaderWriterResponse((HttpServletRequest) request, (HttpServletResponse) response);
+            try {
+                chain.doFilter(request, headerWriterResponse);
+            }
+            finally {
+                headerWriterResponse.writeHeader();
+            }
+            return;
+        }
+        chain.doFilter(request, response);
+    }
+
+    @Override
+    public void destroy() {
+        // intentionally kept empty
+    }
+
+    /**
+     * Response wrapper for writing the Content-Security-Policy header before response is committed
+     */
+    class HeaderWriterResponse extends OnCommittedResponseWrapper {
+
+        private final HttpServletRequest request;
+
+        HeaderWriterResponse(HttpServletRequest request, HttpServletResponse response) {
+            super(response);
+            this.request = request;
+            configureResponse(10485760); // 10 MiB
+        }
+
+        /**
+         * Content-Security-Policy header cannot be set before collection of nonce/hash information in {@link CspScriptsResponseWriter}.
+         * Therefore as a workaround we have to increase the response buffer size.
+         * Otherwise, sending headers after the first flush occurred will not have any effects.
+         * See {@link javax.servlet.ServletResponse#setBufferSize(int)}.
+         * @param bufferSize the buffer size to use
+         */
+        private void configureResponse(int bufferSize) {
+            setBufferSize(bufferSize);
```

Comment:
>  find better solution than increasing the buffer size

@melloware This is some kind of hack that the open task addresses.

Reply:
Yep I see what you are doing here.  This is a tricky one I can see why you solved it this way